src/aiq/observability/register.py

```diff
@@ -26,6 +33,81 @@
 logger = logging.getLogger(__name__)
 
 
+def get_wandb_api_key(config_api_key: Optional[str] = None) -> Optional[str]:
+    """
+    Get the W&B API key from various sources in order of priority:
+    1. Config provided key
+    2. WANDB_API_KEY environment variable
+    Returns:
+        The API key if found, None otherwise
+    """
+    if config_api_key:
+        return config_api_key
+    # Check environment variable
+    env_api_key = os.environ.get("WANDB_API_KEY")
+    if env_api_key:
+        return env_api_key
+    return None
+
+
+class WeaveTelemetryExporter(TelemetryExporterBaseConfig, name="weave"):
+    """A telemetry exporter to transmit traces to Weights & Biases Weave using OpenTelemetry."""
+    entity: str = Field(description="The W&B entity/organization.")
+    project: str = Field(description="The W&B project name.")
+    log_otel_only: bool = Field(
+        default=True,
+        description="If true, only log opentelemetry traces to Weave. If false, log all traces to Weave."
+    )
+    api_key: Optional[str] = Field(
+        default=None,
+        description="Your W&B API key for authentication. If not provided, will look for WANDB_API_KEY environment variable."
+    )
+    endpoint: Optional[str] = Field(
+        default="https://trace.wandb.ai/otel/v1/traces",
+        description="The Weave OTEL endpoint to export telemetry traces. If not provided, will use the default Weave endpoint."
+    )
+
+
+@register_telemetry_exporter(config_type=WeaveTelemetryExporter)
+async def weave_telemetry_exporter(config: WeaveTelemetryExporter, builder: Builder):
+    import base64
+    from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
+
+    # Initialize Weave client if log_otel_only is false
+    # it captures traces from the underlying frameworks weave has integration with
+    # list of weave integrations: https://weave-docs.wandb.ai/guides/integrations/
+    if not config.log_otel_only:
+        import weave
+        _ = weave.init(project_name=f"{config.entity}/{config.project}")
+
+    class NoOpSpanExporter:
+        def export(self, spans):
+            return None
+
+        def shutdown(self):
+            return None
+
+    api_key = get_wandb_api_key(config.api_key)
+    if not api_key:
+        logger.error("W&B API key not found. Please provide it in the config or set WANDB_API_KEY environment variable.")
+        yield NoOpSpanExporter()
+        return
+
+    try:
+        auth = base64.b64encode(f"api:{api_key}".encode()).decode()
```

Comment:
Out of curiosity, whats the purpose for the encode/decode?

Reply:
not needed anymore.